insights/parsers/audit_log.py

```diff
@@ -0,0 +1,144 @@
+            * Note, if a line isn't parseable by the above steps, return dict with items 'raw_message' and 'is_valid' only.
+
+        Parameters:
+            line(string): Log line to parse
+
+        Yields:
+            Dictionary corresponding to the data in line split into
+            functional elements:
+
+            * **raw_message** - the original unparsed line
+            * **is_valid** - true if the line is parseable, else false
+            * **type** - the log type
+            * **timestamp** - the timestamp of the log line (as a string)
+            * **msg_ID** - the serial number behind the timestamp
+            * **unparsed** - the unparsed part in log, may or maynot exist
+
+            If the line is not regular, for some reason, then as many fields as
+            possible are pulled from the line.
+        """
+        info = {}
+        info_basic = {'raw_message': line,
+                      'is_valid': False}
+        linesp = shlex.split(line)
+
+        if (len(linesp) < 2 or
+                not (linesp[0] and linesp[0].startswith('type=')) or
+                not (linesp[1] and linesp[1].startswith('msg=audit('))):
+            return info_basic
+
+        info['type'] = linesp[0].lstrip('type=')
+
+        timestamp_id = linesp[1].lstrip('msg=audit(').rstrip('):')
+        timestamp_id_sp = timestamp_id.split(':')
+        if len(timestamp_id_sp) != 2:
+            return info_basic
+        info['timestamp'] = timestamp_id_sp[0]
+        info['msg_ID'] = timestamp_id_sp[1]
+
+        for index in xrange(len(linesp) - 1, 1, -1):
+            itemsp = linesp[index].split('=', 1)
+            if len(itemsp) < 2:
+                unparsed1 = (line.rsplit(linesp[index + 1])[0]
+                             if index + 1 < len(linesp)
+                             else line)
+                unparsed2 = unparsed1.split(linesp[1])[-1]
+                info["unparsed"] = unparsed2.strip()
+                break
+            info[itemsp[0]] = itemsp[1]
+
+        info.update(info_basic)
+        info['is_valid'] = True
+        return info
+
+    def get_after(self, timestamp, s=None):
+        """
+        Find all the (available) logs that are after the given time stamp.
+        Override this function in class LogFileOutput.
+
+        Parameters:
+
+            timestamp(datetime.datetime): lines before this time are ignored.
+            s(str or list): one or more strings to search for.
+                If not supplied, all available lines are searched.
+
+        Yields:
+            (dict): the parsed data of lines with timestamps after this date in the
+            same format they were supplied.
+        """
+        if isinstance(s, str):
+            invalid_search_result = lambda l: s not in l  # NOQA
+        elif isinstance(s, list):
+            invalid_search_result = lambda l: any(w not in l for w in s)  # NOQA
+        else:
+            invalid_search_result = lambda l: True  # NOQA
```

Comment:
As we discussed, 
- better raise a "TypeError" exception when its type is not `str` or `list`.
- better check the items in the `list`, if there are any non-string items, raise a "TypeError" exception.